src/goal.rs

```diff
@@ -785,11 +786,69 @@ impl<'a> Goal<'a> {
   }
 
   /// Consume this goal and add its case splits to the proof state
-  fn case_split(mut self, state: &mut ProofState<'a>) {
+  fn case_split(mut self, state: &mut ProofState<'a>) -> CaseSplitStatus {
     let new_lemmas = self.add_lemma_rewrites(state);
 
+    let mut blocking_vars: HashSet<_> = HashSet::default();
```

Comment:
Can this whole thing be a separate function? Something like `find_blocking_variables(&self) -> HashSet<Var>`? And can we have a function that returns the next scrutinee (something like `next_scrutinee(&self) -> Option<Var>`) which uses `find_blocking_variables` to pop the first scrutinee that is blocking?

I think we can avoid the `CaseSplitStatus` struct entirely by changing the main loop of `prove` to be along the lines of.

```
if let Some(scrutinee) = goal.next_scrutinee() {
  // note that we now will pass the scrutinee to the case split instead of popping it in the case split.
  goal.case_split(scrutinee, state)
} else {
  // goal is invalid
}
```

Reply:
Yeah I can do that. I didn't really like having the CaseSplitStatus thing in there either.